### PR TITLE
Reworked validation of Modbus tabs

### PR DESCRIPTION
**Pull request by maxunbearable**
## Pull Request description

Reworked validation of Modbus tabs

To save the configuration, the following options are available:
- The server section is filled in
- The master section is filled in (at least 1)
- Both the master and server sections are filled 

## Walkthrough
[Screencast from 30.07.24 12:12:43.webm](https://github.com/user-attachments/assets/849df77a-b8b0-4ba0-a57a-2e02f698aaac)

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [ ] Description contains human-readable scope of changes.
- [ ] Description contains brief notes about what needs to be added to the documentation.
- [ ] No merge conflicts, commented blocks of code, code formatting issues.
- [ ] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts [hunk 1]
```diff
@@ -108,9 +109,15 @@ export class ModbusBasicConfigComponent implements ControlValueAccessor, Validat
...
   }
 
-  validate(): ValidationErrors | null {
-    return this.basicFormGroup.valid ? null : {
-      basicFormGroup: {valid: false}
-    };
+  validate(basicConfigControl: UntypedFormControl): ValidationErrors | null {
+    const masterHasSlaves = !!basicConfigControl.value.master?.slaves?.length;
+    const slaveEnabled = basicConfigControl.value.slave?.sendDataToThingsBoard;
+    const slaveIsValid = this.basicFormGroup.get('slave').valid;
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts:115]: Try using only `basicConfigControl`

**Review by vvlladd28 — CHANGES_REQUESTED** [review]

File: ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts (hunk 1 above)
Comment (maxunbearable) [reply ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts:127]: I think it's not possible to check validity of slave without this.basicFormGroup

File: ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts [hunk 2]
```diff
@@ -77,13 +81,21 @@ export class ModbusBasicConfigComponent implements ControlValueAccessor, Validat
...
       .subscribe(value => {
         this.onChange(value);
         this.onTouched();
       });
+
+    this.enableSlaveControl.valueChanges
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(enable => {
+        this.updateSlaveEnabling(enable);
+        this.basicFormGroup.get('slave').updateValueAndValidity({emitEvent: !!this.onChange});
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts:97]: This method is not required when using enabled/disabled form controls

File: ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-master-table/modbus-master-table.component.ts [hunk 3]
```diff
@@ -138,12 +130,6 @@ export class ModbusMasterTableComponent implements ControlValueAccessor, Validat
     this.pushDataAsFormArrays(master.slaves);
   }
 
-  validate(): ValidationErrors | null {
-    return this.slaves.controls.length ? null : {
-      slavesFormGroup: {valid: false}
-    };
-  }
```
Comment (vvlladd28) [inline ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-master-table/modbus-master-table.component.ts:145]: Why remove validation?

**Review by vvlladd28 — CHANGES_REQUESTED** [review]

File: ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts (hunk 2 above)
Comment (maxunbearable) [reply ui-ngx/src/app/modules/home/components/widget/lib/gateway/connectors-configuration/modbus/modbus-basic-config/modbus-basic-config.component.ts:96]: it's needed to recheck validity on enable button change

**Review by vvlladd28 — APPROVED** [review]

**Pull request merged**